radicle-surf/src/vcs/git/repo.rs

```
@@ -35,23 +34,74 @@ use crate::{
     },
 };
 use directory::Directory;
-use nonempty::NonEmpty;
 use radicle_git_ext::Oid;
 use std::{
     collections::{BTreeSet, HashSet},
     convert::TryFrom,
     str,
 };
 
-/// This is for flagging to the `file_history` function that it should
-/// stop at the first (i.e. Last) commit it finds for a file.
-pub(super) enum CommitHistory {
-    _Full,
-    Last,
+/// An iterator that produces the history of commits for a given `head`,
+/// in the `repo`.
+///
+/// As we have the `head`, this history is guaranteed not empty.
+pub struct History<'a> {
+    repo: RepositoryRef<'a>,
+    head: Commit,
+    revwalk: git2::Revwalk<'a>,
+}
+
+impl<'a> History<'a> {
+    /// Returns the first commit (i.e. the head) in the history.
+    pub fn head(&self) -> &Commit {
+        &self.head
+    }
+
+    /// Returns an iterator that only iterates commits for a file specified by
+    /// `path`, in this history.
+    pub fn has_file(self, path: file_system::Path) -> impl Iterator<Item = Commit> + 'a {
```

Perhaps this should be `Result<Commit, Error>` so that errors aren't filtered out, but instead returned.

---

I wanted to have the `Item` type consistent as the `History` iterator item type, i.e.` Commit`. I agree that this is a trade-off. If we do want to bubble up the potential errors, I think we should change `History` iterator returning `Result<>` as well.

---

Agreed, the iterator only returned `Commit` before it was safe to do so. Don't be afraid to change the code in these sort of cases as semantics change :)

---

I've changed this method to return `History` as well.